monitoring/queuemon.go

```diff
@@ -0,0 +1,204 @@
+	}()
+}
+
+// End closes the watcher
+// As of now this isn't being called outside of tests, as don't have any kind of signal catch/shutdown in the shipper itself.
+func (mon QueueMonitor) End() {
+	if mon.bypass {
+		return
+	}
+	mon.done <- struct{}{}
+}
+
+// updateMetrics is responsible for fetching the metrics from the queue, calculating whatever it needs to, and sending the complete events to the output
+func (mon *QueueMonitor) updateMetrics() error {
+	raw, err := mon.queue.Metrics()
+	if err != nil {
+		return fmt.Errorf("error fetching queue Metrics: %w", err)
+	}
+
+	count, limit, queueIsFull, err := getLimits(raw)
+	if err != nil {
+		return fmt.Errorf("could not get limits: %w", err)
+	}
+
+	if queueIsFull {
+		mon.queueLimitCount = mon.queueLimitCount + 1
+	}
+
+	mon.sendToOutputs(reporter.QueueMetrics{
+		CurrentQueueLevel:      opt.UintWith(count),
+		QueueMaxLevel:          opt.UintWith(limit),
+		QueueIsCurrentlyFull:   queueIsFull,
+		QueueLimitReachedCount: opt.UintWith(mon.queueLimitCount),
+		// Running on a philosophy that the outputs should be dumb and unopinionated,
+		//so we're doing the type conversion here.
+		OldestActiveTimestamp: raw.OldestActiveTimestamp.String(),
+	})
+
+	return nil
+}
+
+func (mon QueueMonitor) sendToOutputs(metrics reporter.QueueMetrics) {
+	for _, out := range mon.outputs {
+		err := out.ReportQueueMetrics(metrics)
+		//Assuming we don't want to make this a hard error, since one broken output doesn't mean they're all broken.
+		if err != nil {
+			mon.log.Errorf("Error sending to output: %w", err)
+		}
+	}
+
+}
+
+// Load the raw config and look for monitoring outputs to initialize.
+func initOutputs(cfg Config) []reporter.Reporter {
+	outputList := []reporter.Reporter{}
+
+	if cfg.Outputs.LogOutput.Enabled {
+		reporter := log.NewLoggerReporter()
+		outputList = append(outputList, reporter)
+	}
+
+	if cfg.Outputs.ExpvarOutput.Enabled {
+		reporter := expvar.NewExpvarReporter(cfg.Outputs.ExpvarOutput)
+		outputList = append(outputList, reporter)
+	}
+
+	return outputList
+}
+
+// This is a wrapper to deal with the multiple queue metric "types",
+// as we could either be dealing with event counts, or bytes.
+// The reporting interfaces assumes we only want one.
+func getLimits(raw outqueue.Metrics) (uint64, uint64, bool, error) {
+
+	//bias towards byte count, as it's a little more granular.
+	if raw.ByteCount.Exists() && raw.ByteLimit.Exists() {
+		count := raw.ByteCount.ValueOr(0)
+		limit := raw.ByteLimit.ValueOr(0)
+		// As @faec has noted, calculating limits can be a bit awkward when we're dealing with reporting/configuration in bytes.
+		//All we have now is total queue size in bytes, which doesn't tell us how many more events could fit before we hit the queue limit.
+		//So until we have something better, mark anything as 90% or more full as "full"
+
+		// I'm assuming that limit can be zero here, as perhaps a user can configure a queue without a limit, and it gets passed down to us.
```

Comment:
I'm wondering if it is misleading to track this here. It seems like the queue should just report its current fill level, and we should track the count of events that were rejected because they could not fit on the publisher side (likely the RPC handlers).

Reply:
@faec what do you think? Does it make more sense to track queue levels here, or in the shipper side? My thinking was to keep the shipper `Metrics()` API more low-level, but I'm kinda on the fence.

Reply:
Well, I do like the idea of reporting the explicit levels without trying to summarize into "full" or "not full" since that's always going to be somewhat brittle.

I'm not sure what the right alternative would be, though. For example, for a queue configured to block rather than rejecting when full, the important question might be how many shipper requests blocked for a nontrivial duration, or the total amount of time consumers spent waiting for queue space over the last [time interval].

I'm not sure we know enough to pick the right summary fields yet, though. I expect we'll have more basis for deciding as we develop standard benchmarks. So from that perspective, I'd be fine keeping a "full" field that just checks a 90% threshold, and leaving a comment saying that this is a placeholder to show off the monitoring framework and should be replaced by a more principled summary once we've collected enough data.

Reply:
Yah, agreed that a lot of these summary fields are definitely gonna change over time. I'm thinking we should be comfortable with expanding/tweaking these metrics over the next few releases. A lot of these metrics might make more/less sense once we have the actual queues and everything hooked up too.

Reply:
Makes sense to me. The obvious place to measure this will likely show up as we build more of the shipper.